### ISSUE TITLE
Offset HUD Ladder

### ISSUE THREAD

**Issue by DeputyLOL** [Mihara/RasterPropMonitor#418]
I know that the HUD ladder is always at the centre of the screen but is it possible to nudge it left or right?

![image](https://cloud.githubusercontent.com/assets/12616126/10728733/43ad14a2-7bdc-11e5-93e7-76b849575dfe.png)

I'm modelling my PFD off of a B737 NG. Problem is, I cant fit a very crucial bar in as there isn't enough space. The vertical speed bar and altitude bar are on the same side, the ladder is slightly offset to the left as the airspeed bar is the only bar on the left.

Please let me know if this is possible.
(also, how are the sliding numbers for screens coming along? #365)

**Comment by MOARdV** [comment]
Yes, it looks like the HUD ladder could be changed to have its center point defined as a config variable.  As for #365, I haven't started on it.  I am slowing down my modding pace, since work has been busy and I don't want to burn out on coding during my off-work hours.

**Comment by DeputyLOL** [comment]
fair enough. enjoy your off-work hours.

**Comment by DeputyLOL** [comment]
Do you think you could add the ability to have an 'enable ladder' config like with the vertical bars?

**Comment by MOARdV** [comment]
> Do you think you could add the ability to have an 'enable ladder' config like with the vertical bars?

So it could be switched off?  Off-hand, I think that would be fairly easy to implement.

**Comment by DeputyLOL** [comment]
Cool, I'm start to make 'random failures' a thing in my cockpit. Every now and then a circuit breaker will pop and a certain system would be off line until you reset the braker.

**Comment by DeputyLOL** [comment]
I have a question about the acceleration variable. Can that only be positive or can it measure deceleration as well?

**Comment by MOARdV** [comment]
> I have a question about the acceleration variable. Can that only be positive or can it measure deceleration as well?

Which acceleration variable?  `ACCEL` measures current thrust over mass, so it's always positive (since thrust and mass can never be negative).  `EFFECTIVEACCEL` is a KSP-derived value, and it measures the magnitude of a thrust vector (thus always positive).

**Comment by DeputyLOL** [comment]
I think I'm referring to EFFECTIVEACCEL. The problem I have with it is that it is always positive. Could we have a variable that is + when accelerating and - when decelerating?

**Comment by DeputyLOL** [comment]
Thank you for setting up request.
I'm trying to make speed trend arrows (green arrow you see pointing up the speed display). It tells you the speed you will reach in 10 seconds with current acceleration. I can do all the equations needed, I just need the acceleration. Does that give me the ground speed acceleration or airspeed acceleration?

**Comment by MOARdV** [comment]
> I'm trying to make speed trend arrows (green arrow you see pointing up the speed display). It tells you the speed you will reach in 10 seconds with current acceleration. I can do all the equations needed, I just need the acceleration. Does that give me the ground speed acceleration or airspeed acceleration?

Let's keep discussion of the needs of that variable with the new issue, so the discussion doesn't get lost.

**Comment by DeputyLOL** [comment]
ok

**Comment by MOARdV** [comment]
Okay, I'm finally getting this done.  Sorry it took so long.

**Comment by DeputyLOL** [comment]
WOOOOOOOO!!!!!!!!!!!

**Comment by MOARdV** [comment]
Available in the current [DropBox](https://www.dropbox.com/s/qocogwy8qm1qxi2/RasterPropMonitor-dev.0.26.1.zip?dl=0) dev build, by the way.

**Comment by DeputyLOL** [comment]
I know, im trying it right now.

**Comment by DeputyLOL** [comment]
> > Do you think you could add the ability to have an 'enable ladder' config like with the vertical bars?

> So it could be switched off? Off-hand, I think that would be fairly easy to implement.

Did you get to the enable/disable ladder feature? :D

**Comment by DeputyLOL** [comment]
Doesn't work for me. I have this;
horizonOffset = -400,-15
And the ladder is still in the exact same place as it was.

**Comment by DeputyLOL** [comment]
The offset works but it doesn't affect the edges of the ladder. 
The centre of the ladder will move but the edges will not.

**Comment by MOARdV** [comment]
What do you mean by the edges?  When I tested it, the ladder moved (the area where the ladder is, and the cropping box that contains it).  The vertical strips (altitude, speed) are separate components, so they need to be moved, too.  Or is it something else I missed (entirely possible, since I don't use a HUD very often).

**Comment by DeputyLOL** [comment]
The cropping box doesn't seem to be moving for me. I'm still testing it now.

**Comment by DeputyLOL** [comment]
![monitor2016-05-09_16-00-30_-1549964_60](https://cloud.githubusercontent.com/assets/12616126/15117400/49aec6f8-15ff-11e6-8170-7b6b784215b4.png)

It might be a little tricky to tell (because its partly still in transition) but the ladder are not symmetrical.
You see the left side stops just before the aircraft symbol and the right one continues off the edge of the indicator. The ladder is centred but the middle of it is offset.

**Comment by MOARdV** [comment]
Okay.  Very strange.  Probably the problem I had is that I have a very simply HUD ladder to work with (the one that comes with stock RPM).  Is your ladder texture available in your cockpit mod?  Or can you put it somewhere I can download for testing?

**Comment by DeputyLOL** [comment]
it is in the cockpit mod.
GameData\AdvancedCockpit\Displays\Textures\PFD_ladder.png

**Comment by DeputyLOL** [comment]
![pfd_ladder](https://cloud.githubusercontent.com/assets/12616126/15120696/d911923a-160e-11e6-8432-56824e367c09.png)

But i can put it here and you can copy it from here

**Comment by MOARdV** [comment]
I don't see the problem when I use this ladder on the stock HUD.  The center of the ladder shifts over to the side with the rest of the texture.  What are you using for the config values for the ladder?

**Comment by DeputyLOL** [comment]
horizonTexture = AdvancedCockpit/Displays/Textures/PFD_ladder
			horizonOffset = -45,15
			use360horizon = false
			horizonSize = 299,309
			horizonTextureSize = 800,800

**Comment by MOARdV** [comment]
I took a look at it with your settings, and I still do not see anything unusual.  The ladder is moved to the right a little ways, and I don't see anything else amiss.  I double-checked, and the cropping shader is updating correctly.

**Comment by DeputyLOL** [comment]
seems strange. ill do more testing on my end to see if I missed something.

**Comment by DeputyLOL** [comment]
What do you think might be causing it?

**Comment by DeputyLOL** [comment]
I've just started testing. I have an idea about the cause. I have another separate monitor that doesn't have an offset ladder yet, do you think that RPM is getting the ladder centre from one monitor config and then confusing the cropping from the other monitor?

**Comment by DeputyLOL** [comment]
No luck.
![monitor2016-05-19_12-54-02_-715498_59](https://cloud.githubusercontent.com/assets/12616126/15392787/4d3d96c2-1dc1-11e6-986f-8d5b803e3f07.png)
![monitor2016-05-19_12-52-00_-715498_59](https://cloud.githubusercontent.com/assets/12616126/15392788/4d3eb3ea-1dc1-11e6-9455-f05a010af1b5.png)

**Comment by MOARdV** [comment]
Oh, wait... you have more than one HUD monitor with different configurations?  It might be because I set the crop bounds shader parameters once at start up, and maybe the second monitor (with no offset) is overwriting the first one's shader parameters.  Let me make a change.

**Comment by DeputyLOL** [comment]
I'm sure you'll figure it out, let me know when you want me to test it. :D

**Comment by MOARdV** [comment]
Okay.  I think [this](https://www.dropbox.com/s/qocogwy8qm1qxi2/RasterPropMonitor-dev.0.26.1.zip?dl=0) will do the trick.

**Comment by DeputyLOL** [comment]
No change sorry

**Comment by DeputyLOL** [comment]
I going to keep testing.
I've tried putting hudOffset = 0,0 onto the displays that don't have an offset HUD, that didn't work.
I've tried removing the monitor that doesn't have an offset hud from the IVA, that didn't work.
I've tried removing the monitor that doesn't have an offset hud from the mod completely, that didn't work.
I'm now out of ideas...

**Comment by MOARdV** [comment]
Can you put a ZIP of your IVA (or a simpler IVA with the appropriate MFDs) somewhere I can download?  I am not seeing a problem here, so I don't know what is going on.

**Comment by DeputyLOL** [comment]
Dev Version: https://www.dropbox.com/s/lxs398gjlyuaws3/KSP_AdvancedCockpit_DEV001.zip?dl=0
Put the "AdvancedCockpit" folder into GameData.
To avoid having to go through the startup procedure I have put in a dev switch. This should make the plane think it has power when it doesn't. But.... its untested so I'm not sure it will work. If it doesn't I included a walkthrough to starting the plane up. I hope this helps.

There are 3 separate monitor configs with HUDs. 2 of them are identical.
These can be found in AdvancedCockpit/Displays/MainInstrumentPanel/...
PFD_MIP_CAP_Display
PFD_MIP_FO_Display
ISFD_MIP_Display

**Comment by DeputyLOL** [comment]
Was this fix implemented in the latest dev build?

**Comment by MOARdV** [comment]
I have not reproduced the problem here yet.  I will try to work on it this weekend.

**Comment by DeputyLOL** [comment]
Thank you.

**Comment by DeputyLOL** [comment]
![horizon offset](https://cloud.githubusercontent.com/assets/12616126/15749604/e1b5ef24-28da-11e6-9e28-e02ad57dd526.png)

I've been digging deeper into this and I may have found something that should help you pin point the cause of my problem. 
I noticed that even though the cropbox wasn't where it should be, it didn't stay in the centre either which meant it was offset as-well. I put a screenshot of my HUD into paint and did some measuring.
I also labelled on the screenshot the centre of the texture, centre of the ladder and centre of the cropbox. BUT WE MUST LOOK DEEPER! (ignore the '2069')

When the Offset HUD feature was first impletemented I noticed that something was off with the y-coordinate. I always use that 'y being positive = UP', but that's because I'm from the UK. I'm assuming that you use the 'y being positive = DOWN' as that is how everything else works. :) 
The exception to that is the horizonOffset. In the texture above you can see my config and the results of it. 

Long story short;
The y axis is reversed.
The cropbox offset is half of the ladder offset and the y axis is reversed... again.

**Comment by DeputyLOL** [comment]
Does this help you?

**Comment by MOARdV** [comment]
> Does this help you?

It does.  I will get back to this once I get the docking / undocking problems sorted out.  They have consumed all of my time this week.

**Comment by DeputyLOL** [comment]
nice. hope you get it sorted.

quick question. I recall you saying that the monitor gauges would be very hard to implement. So instead of making something that functions like a round version of the vertical and horizontal bars, do you think it would be possible to control a texture using the same kind of methods that we can use for props with JSIVariableAnimator? E.g. transform, rotate, scale, maybe even colour?. The only real difference would be that it is 2D and it uses a texture rather than a prop. I don't know how hard that would be to implement.

**Comment by MOARdV** [comment]
A texture on the screen?  That'd be new functionality (that is, we don't currently), although I am pretty sure alexustas has expressed interest in something similar as well.  I suspect it wouldn't be hard to do, really..

**Comment by DeputyLOL** [comment]
Noice!

**Comment by Alexustas** [comment]
> although I am pretty sure alexustas has expressed interest in something similar as well.

I'm very interested in implementing [this](https://github.com/Mihara/RasterPropMonitor/issues/210, and [this ](https://github.com/Mihara/RasterPropMonitor/issues/445)requests. This functionality will cover all my needs, I'm sure ))

**Comment by MOARdV** [comment]
By the way, @DeputyLOL , thank you for your sleuthing on the offsets.  I *think* [this update](https://www.dropbox.com/s/qocogwy8qm1qxi2/RasterPropMonitor-dev.0.26.1.zip?dl=0) has it fixed (no promises - this evening is a bit chaotic).

**Comment by DeputyLOL** [comment]
thank you, ill test it now. I'm assuming that the y-offset will have to be -15 now.
You should have seen me when I started figuring it out, I felt like a detective. :D

**Comment by DeputyLOL** [comment]
from the 4 seconds of gameplay I got before my craft magically disappeared, I have to report that there wasn't any change. I'm gonna keep testing and see if I missed something.

**Comment by DeputyLOL** [comment]
Actually look like it got worse. :(

**Comment by DeputyLOL** [comment]
![horizon differences](https://cloud.githubusercontent.com/assets/12616126/15962253/98d93f54-2f01-11e6-826f-782ef4997af4.png)

Looks like the crop box is even more offset than it was previously. I'm gonna put my test hud back to get the differences.

Also the y offset is still backwards.

**Comment by DeputyLOL** [comment]
Ok so I did some more detectiveing and I came up empty. There isn't a clear pattern like in the diagram before.
![horizon offset2](https://cloud.githubusercontent.com/assets/12616126/15962988/9e233e9c-2f06-11e6-97b2-418762d85e15.png)
The y offset is still reversed, the cropbox y offset is 3 times than the ladder offset (and is reversed) and the cropbox x offset is 3.6' smaller than the ladder offset. Hope this helps.

**Comment by MOARdV** [comment]
Okay.  I've already gone derper here.  It's been too long since I looked at the HUD code, and I had way too many distractions yesterday afternoon.  I'll pull this back apart and put it together again, and I'll put together a texture that has grid lines so I can read the effects of the changes more clearly.

As for Y offsets - I'm using the screen-space convention of (0, 0) being the upper-left corner, with a left-handed coordinate system, so Y+ is down.  Yes, it's wrong from conventional Euclidean standards, but I thought it would be more approachable for people who aren't familiar with working in 3-space.  It also follows the convention of identifying rows of text, since we read top-to-bottom on a page.  It also should be the convention that the rest of the parameters in the HUD are using.

Or maybe it's a side-effect of my Yankee habit of driving on the wrong since of the roads. :)

Since the major code-breaking changes are out of the way, I'll focus on this, next.

**Comment by DeputyLOL** [comment]
> As for Y offsets - I'm using the screen-space convention of (0, 0) being the upper-left corner, with a left-handed coordinate system, so Y+ is down. Yes, it's wrong from conventional Euclidean standards, but I thought it would be more approachable for people who aren't familiar with working in 3-space. It also follows the convention of identifying rows of text, since we read top-to-bottom on a page. It also should be the convention that the rest of the parameters in the HUD are using. <

I'm aware of that but you seem to have it the 'right' way around. My config says '+15' (which should be down) and the ladder centre moves '-15'. I suspect that its not suppose to be the European way. :P

**Comment by DeputyLOL** [comment]
I suspect that I may have confused the situation by saying the cropbox y offset is reversed.
I think its just the ladder offset that is reversed.

**Comment by MOARdV** [comment]
Yes, ladder offset was y-reversed, because I forgot that the offsets are in RH coordinates, so I needed to flip the sign for ladder offset.  And the cropbox offset is in normalized screen space coordinates, so I needed to scale the offset there by x2, not by x1/2 like I did last night (I knew what I needed to do, I just inverted it).

I think the update now on [DropBox](https://www.dropbox.com/s/qocogwy8qm1qxi2/RasterPropMonitor-dev.0.26.1.zip?dl=0) is correct - I looked at a simplified texture with reference marks, and both displacements and crop box looked like they moved the correct direction.

**Comment by DeputyLOL** [comment]
Slight improvement. It looks like the x axis is correct now. Yay!
The y axis... not so much.
![monitor2016-06-10_14-48-30_-358872_149](https://cloud.githubusercontent.com/assets/12616126/15966336/a48fa6f8-2f1a-11e6-8d35-0df3aaaa8e12.png)

**Comment by DeputyLOL** [comment]
My guess is that the y axis is reversed some where again. :D

**Comment by DeputyLOL** [comment]
Or it might just be me not reverting the '15' to a '-15'!

**Comment by DeputyLOL** [comment]
Works perfectly now! Thanks you so much.

Before we finish with this though, do you think you could add an 'enable/disable' ladder variable. Pretty please. :D

**Comment by MOARdV** [comment]
> Before we finish with this though, do you think you could add an 'enable/disable' ladder variable.

Similar to the vertical bar / horizontal bar enabling variable?  That should be easy (famous last words).

**Comment by DeputyLOL** [comment]
Yes thank you.

**Comment by MOARdV** [comment]
[DropBox](https://www.dropbox.com/s/qocogwy8qm1qxi2/RasterPropMonitor-dev.0.26.1.zip?dl=0)

* **horizonEnableVariable** -- Optional, a variable that can be used to control when the HUD ladder is rendered.  If not present, the ladder is always rendered.
* **horizonEnableRange** -- Optional, two values (numbers or [[Defined Variables]]).  When the `horizonEnableVariable` is within the range, the HUD ladder is rendered.  If the variable is outside that range, it is not.

**Comment by DeputyLOL** [comment]
Testing now.

**Comment by DeputyLOL** [comment]
It works perfectly! Thank you!
I think this is closable now.
